rpc/src/services/base_services.rs

```diff
@@ -227,8 +298,15 @@ pub(crate) fn get_block_protocols(
 }
 
 /// Extract the current_protocol and the next_protocol from the block metadata
+#[cached(
+    name = "BLOCK_ADDITIONAL_DATA_CACHE",
+    type = "TimedSizedCache<(ChainId, BlockHash), Option<BlockAdditionalData>>",
```

Comment:
@tizoc @mambisi @jurajselep 
I am not sure about cachcing Optional here,
lets say situation:
- we have current head on **5**
- we receive request /chains/main/blocks/**6**/additional_data
-- so we add to cache: None
- then we apply block **6** + we trigger warmup for 6
- we receive another call /chains/main/blocks/**6**/additional_data

what does that second call /chains/main/blocks/**6**/additional_data returns?
I guess None, instead of Some(data)

how does it cache Optional, I guess, we should not cache optional,
I am going to add some tests to drone CI, to cover something like this and see

Reply:
@tizoc @mambisi @jurajselep 
 there should be a cfg for optionals:

![image](https://user-images.githubusercontent.com/8159517/124454869-3a1a6b00-dd89-11eb-81c4-07fe575105e3.png)

Reply:
Its using a Hashmap under the hood so 
```
let mut map = HashMap::new();
map.insert(("Nice".to_owned(), Some("Game".to_string())), 2);
map.insert(("Nice".to_owned(), None), 2);
```

Output:
```
{
    (
        "Nice",
        Some(
            "Game",
        ),
    ): 2,
    (
        "Nice",
        None,
    ): 2,
}
```

Reply:
@tizoc @mambisi  @jurajselep 
hmm, there is no possibility "not to cache Ok(None)", 
I have a simple test for that

Reply:
@bkontur since we can use any structure we want for cache, it should just be a matter of using one that ignores anything that results in `None`.

Reply:
@bkontur btw, if we receive a request for `6`, but `6` has not been processed yet, cannot that be considered an error?

I am not very clear on the logic of how this should work. After `6` gets applied, is the additional data for it fixed forever? or is there anything that could update that data?

I am thinking that maybe caching the `None` is ok, what is actually needed is to remove the value from the cache once block `6` is applied, and before the warmup function is called (that `None` should be a valid result until then anyway).

Reply:
Still trying to figure out the code that `cached` generates for these annotations, but what I have in mind is (roughly) something like this:

```
super::base_services::BLOCK_ADDITIONAL_DATA_CACHE.get_mut().unwrap().cache_remove(&(chain_id.clone(), block.hash.clone()));
```

Reply:
> 
> 
> Still trying to figure out the code that `cached` generates for these annotations, but what I have in mind is (roughly) something like this:
> 
> ```
> super::base_services::BLOCK_ADDITIONAL_DATA_CACHE.get_mut().unwrap().cache_remove(&(chain_id.clone(), block.hash.clone()));
> ```

@tizoc 
I was thinking exactly the same, to access that caches, but it is another coding,
so I refactored `get_adiition_data->Result<Optional`  to `get_additional_data_or_fail<Result<BLockAdditionData`
like you suggested before comment :)

Reply:
@bkontur yes that is my preferred solution, feels more robust than manually invalidating the cache.

Reply:
@tizoc 
and what about things we put into the cache,
shouldnt we wrap them with `Arc` to aviod cloning, or this is not a issue?

Reply:
@bkontur very good point. For small values I guess cloning is not that big of a deal, but some of the values we cache are huge. Unless you did already, I will try adding that.

Reply:
@bkontur actually, `get` methods borrow values, no cloning happens: https://github.com/jaemk/cached/blob/12858a8a4af7f278e32ae90383d7483a54f829c6/src/lib.rs#L475-L479

And caches are wrapped by a `Mutex`, so no extra synchronization is needed.

Reply:
Wait, actually, that is still wrong (some cloning needs to be happening to return the results from the cached version of the function), let me check what the macro does exactly.

Reply:
Ok, it says so here: https://github.com/jaemk/cached/blob/12858a8a4af7f278e32ae90383d7483a54f829c6/src/lib.rs#L248

Adding `Arc` now (for the big ones at least).

Reply:
Just pushed the `Arc` wrappers for most (but not all) of the cached functions.

Reply:
re-pushed, got eaten by your last push.